### DISCUSSION TITLE
Can type-stubs be shipped with this library for supporting type checking with pyright?

### DISCUSSION THREAD

**Discussion by maneetgoyal** [FreeOpcUa/opcua-asyncio#384] [General]
[Type-stubs](https://github.com/microsoft/pyright/blob/master/docs/type-stubs.md)

**Comment by oroulet** [comment]
I have not used pyright, but I am not against typing. So feel free to create a merge request adding typing where it make sense

**Comment by swamper123** [comment]
These type checking topic would be a good usecase for annotations, which may take a while in our repo.

While pyright is written in e.g. in Typescript and is very microsoft orientated (in my private opinion a nono for open source projects), I would recommened [mypy ](https://github.com/python/mypy)as an alternative. It is written in Python and type-stubs seems to use it as well.

I know from another repo, that it is possible to run github actions with [mypy ](https://github.com/python/mypy)as well, but we should be aware, that adding such a test will be causing lot's of work and fails. ^^

**Reply by oroulet** [reply]
I never managed to make mypy catch really broken code. What it catches is mainly typing errors..
but typing is very nice for documentation in python

**Reply by swamper123** [reply]
Well, yes...that is it's purpose. Catching mostly typing errors. ^^

**Reply by oroulet** [reply]
the final goal should be to catch programing errors that make the program fail.... not only typo in typing declarations

**Reply by swamper123** [reply]
Wouldn't it be a programming error if an argument would be from type "str" but should be "int" and the method would still work in some way with it?
Or a return value is not from the type as expected, perhaps causing unknown behaviour?

As well, it would be additionally, not replacing something.

**Reply by maneetgoyal** [reply]
While refactoring a larger source code (say while changing some function signatures slightly), these type checks can act as a good guiding tool in deciding what other code needs to be changed. I mostly use TypeScript for my projects and type-checking has really enhanced my DX.
